### PR TITLE
feat(format): implement format trait

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog

- Improvement

## Related Issues

Fixes #issue

### CODE REVIEW

**Comment by zhang2014** [comment]
Parquet and JSON input formats will be implemented in the next pr.

File: common/datavalues/src/types/deserializations/number.rs [hunk 1]
```diff
@@ -98,8 +102,30 @@ where
...
-    fn de_null(&mut self, _format: &FormatSettings) -> bool {
-        false
+    fn de_text_csv<R: BufferRead>(
+        &mut self,
+        reader: &mut CheckpointReader<R>,
+        settings: &FormatSettings,
+    ) -> Result<()> {
+        let maybe_quote = reader.ignore(|f| f == b'\'' || f == b'"')?;
+
+        if maybe_quote && reader.ignore(|f| f == b'\'' || f == b'"')? && settings.empty_as_default {
```
Comment (sundy-li) [inline common/datavalues/src/types/deserializations/number.rs:112]: if maybe_quote is false, we should respect `empty_as_default`

File: common/io/src/buffer/buffer_read_ext.rs [hunk 2]
```diff
@@ -78,9 +84,21 @@ pub trait BufferReadExt: BufferRead {
 
     fn must_ignore_byte(&mut self, b: u8) -> Result<()> {
         if !self.ignore_byte(b)? {
+            let buf = self.fill_buf()?;
             return Err(std::io::Error::new(
                 ErrorKind::InvalidData,
-                format!("Expected to have char {}", b as char),
+                format!("Expected to have char {}, {}", b as char, buf[0] as char),
```
Comment (sundy-li) [inline common/io/src/buffer/buffer_read_ext.rs:90]: `buf` may be empty

File: common/io/src/buffer/buffer_read_ext.rs [hunk 3]
```diff
@@ -78,9 +84,21 @@ pub trait BufferReadExt: BufferRead {
...
+        }
+        Ok(())
+    }
+
+    fn must_ignore_white_spaces_and_byte(&mut self, b: u8) -> Result<()> {
+        if !self.ignore_white_spaces_and_byte(b)? {
+            let buf = self.fill_buf()?;
+            return Err(std::io::Error::new(
+                ErrorKind::InvalidData,
+                format!("Expected to have char {}, {}", b as char, buf[0] as char),
```
Comment (sundy-li) [inline common/io/src/buffer/buffer_read_ext.rs:101]: ditto

File: common/io/src/buffer/buffer_read_number_ext.rs [hunk 4]
```diff
@@ -66,9 +66,17 @@ where R: BufferRead
             let _ = self.ignores(|f| (b'0'..=b'9').contains(&f))?;
         }
 
-        FromLexical::from_lexical(buf.as_slice()).map_err_to_code(ErrorCode::BadBytes, || {
-            format!("Cannot parse value:{:?} to number type", buf)
-        })
+        match buf.is_empty() {
+            true => Ok(T::default()),
```
Comment (sundy-li) [inline common/io/src/buffer/buffer_read_number_ext.rs:70]: if empty, defaults to return default value?

Comment (sundy-li) [reply common/io/src/buffer/buffer_read_number_ext.rs:70]: there is a setting `empty_as_default`.

Comment (zhang2014) [reply common/io/src/buffer/buffer_read_number_ext.rs:70]: yes. string type have the same behavior.

File: query/src/formats/format_csv.rs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        let mut deserializers = Vec::with_capacity(self.schema.num_fields());
+        for field in self.schema.fields() {
+            let data_type = field.data_type();
+            deserializers.push(data_type.create_deserializer(self.min_accepted_rows));
+        }
+
+        let mut state = std::mem::replace(state, self.create_state());
+        let state = state.as_any().downcast_mut::<CsvInputState>().unwrap();
+        let cursor = Cursor::new(&state.memory);
+        let reader: Box<dyn BufferRead> = Box::new(BufferReader::new(cursor));
```
Comment (sundy-li) [inline query/src/formats/format_csv.rs:188]: Seems we don't need `Box` here

File: query/src/formats/format_csv.rs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+            field_delimiter,
+            need_skip_header,
+            min_accepted_rows,
+            min_accepted_bytes,
+        }))
+    }
+
+    fn find_quotes(buf: &[u8], pos: usize, state: &mut CsvInputState) -> usize {
+        for (index, byte) in buf.iter().enumerate().skip(pos) {
+            if *byte == b'"' {
```
Comment (sundy-li) [inline query/src/formats/format_csv.rs:107]: what if single quota?

**Comment by BohuTANG** [comment]
Oops, this PR is broken after https://github.com/datafuselabs/databend/pull/5233 merged

**Review by sundy-li — APPROVED** [review]

**Comment by BohuTANG** [comment]
@mergify update

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**